### PR TITLE
[B + C] Adds a villager trading API. Adds BUKKIT-2138, BUKKIT-3935

### PR DESCRIPTION
**Pull request by thefishlive**
The Issue:

Currently in bukkit there is no way to get or set merchant trades or get a merchant from its inventory

PR Breakdown:

This PR adds a way to get, set and add merchant trades and adds a way to get the merchant from the MerchantInventory

Testing Results and Materials:

https://github.com/thefishlive/Bukkit-test - source
https://www.dropbox.com/sh/9v7cy8c2acsuywn/mqdvl8kL3S/Bukkit-Test-0.0.1-SNAPSHOT.jar - binary

Relevant PR(s):

CB-1111 -https://github.com/Bukkit/CraftBukkit/pull/1111 - implmentation

JIRA Ticket:

https://bukkit.atlassian.net/browse/BUKKIT-2138
https://bukkit.atlassian.net/browse/BUKKIT-3935

### CODE REVIEW

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public List<MerchantTrade> getTrades(Player player);
+
+    /**
+     * Add a trade to the merchants trades
+     * 
+     * @param trade a new trade for the villager to perform
+     */
+    public void addTrade(MerchantTrade trade);
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public void addTrade(MerchantTrade trade);
+    
+    /**
+     * Remove a trade from the merchants trades
+     * 
+     * @param trade the trade to remove
+     */
+    public void removeTrade(MerchantTrade trade);
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public void removeTrade(MerchantTrade trade);
+    
+    /**
+     * Set the list of trades for this merchant
+     * 
+     * @param trades the new trade list
+     */
+    public void setTrades(List<MerchantTrade> trades);
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    
+    /**
+     * Represents a trade a villager can make
+     */
+    public class MerchantTrade {
+
+    	private ItemStack product;
+    	private ItemStack[] price;
+		private int tradesLeft;
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    	
+    	/**
+    	 * Create a new villager trade defining the product and price for that product
+    	 * 
+    	 * @param product the item that the villager sells
+    	 * @param price the item(s) that the villager defines as the price for the product
+    	 */
+    	public MerchantTrade(ItemStack product, ItemStack...  price) {
+    		Validate.isTrue(price.length <= 2, "The price can only be a maximum of two items");
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    	private ItemStack[] price;
+		private int tradesLeft;
+    	
+    	/**
+    	 * Create a new villager trade defining the product and price for that product
+    	 * 
+    	 * @param product the item that the villager sells
+    	 * @param price the item(s) that the villager defines as the price for the product
+    	 */
+    	public MerchantTrade(ItemStack product, ItemStack...  price) {
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Needs a `throws IllegalArgumentException` in the Javadoc and a `throws` on the end of the method

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    		Validate.isTrue(price.length <= 2, "The price can only be a maximum of two items");
+    		
+    		this.product = product;
+    		this.price = price;
+    	}
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    	
+    	/**
+    	 * Get the price of this trade
+    	 * 
+    	 * @return the price of this trade, either one or two item stacks in array
+    	 */
+    	public ItemStack[] getPrice() {
+    		return price;
+    	}
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    	
+    	/**
+    	 * Get the product this villager trades for
+    	 * 
+    	 * @return the product that this villager trades
+    	 */
+    	public ItemStack getProduct() {
+    		return product;
+    	}
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    	}
+    	
+    	/**
+    	 * Set the price of this trade
+    	 * 
+    	 * @param price the price of the trade, a maximum of two items
+    	 */
+    	public void setPrice(ItemStack... price) {
+    		Validate.isTrue(price.length <= 2, "The price can only be a maximum of two items");
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    	 * Set the price of this trade
+    	 * 
+    	 * @param price the price of the trade, a maximum of two items
+    	 */
+    	public void setPrice(ItemStack... price) {
+    		Validate.isTrue(price.length <= 2, "The price can only be a maximum of two items");
+    		
+    		this.price = price;
+    	}
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    	
+    	/**
+    	 * Set the product that this villager trades
+    	 * 
+    	 * @param product the product the villager trades
+    	 */
+    	public void setProduct(ItemStack product) {
+    		this.product = product;
+    	}
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    	
+    	/**
+    	 * Gets the amount of trades this specific trade has left
+    	 * 
+    	 * @return the amount of trades left
+    	 */
+    	public int getTradesLeft() {
+    		return tradesLeft;
+    	}
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+     * @param trades the new trade list
+     */
+    public void setTrades(List<MerchantTrade> trades);
+    
+    /**
+     * Represents a trade a villager can make
+     */
+    public class MerchantTrade {
+
+    	private ItemStack product;
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Formatting is odd here. Also, why isn't MerchantTrade in a new file?

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * Represents a Entity that can trade with other HumanEntities
+ */
+public interface Merchant extends Entity {
+
+
+    /**
+     * Get a list of trades that the villages trades
+     *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * Get a list of trades that the villages trades
+     * 
+     * @return all the trades the villager can perform
+     */
+    public List<MerchantTrade> getTrades();
+
+    /**
+     * Get a list of trades that the villages trades for a specific player
+     *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * Get a list of trades that the villages trades for a specific player
+     * 
+     * @return all the trades the villager can perform to a specific player
+     */
+    public List<MerchantTrade> getTrades(Player player);
+
+    /**
+     * Add a trade to the merchants trades
+     *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * Add a trade to the merchants trades
+     * 
+     * @param trade a new trade for the villager to perform
+     */
+    public void addTrade(MerchantTrade trade);
+    
+    /**
+     * Remove a trade from the merchants trades
+     *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * Remove a trade from the merchants trades
+     * 
+     * @param trade the trade to remove
+     */
+    public void removeTrade(MerchantTrade trade);
+    
+    /**
+     * Set the list of trades for this merchant
+     *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public class MerchantTrade {
+
+    	private ItemStack product;
+    	private ItemStack[] price;
+		private int tradesLeft;
+    	
+    	/**
+    	 * Create a new villager trade defining the product and price for that product
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    	public MerchantTrade(ItemStack product, ItemStack...  price) {
+    		Validate.isTrue(price.length <= 2, "The price can only be a maximum of two items");
+    		
+    		this.product = product;
+    		this.price = price;
+    	}
+    	
+    	/**
+    	 * Get the price of this trade
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+    	 * 
+    	 * @return the price of this trade, either one or two item stacks in array
+    	 */
+    	public ItemStack[] getPrice() {
+    		return price;
+    	}
+    	
+    	/**
+    	 * Get the product this villager trades for
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    	 * 
+    	 * @return the product that this villager trades
+    	 */
+    	public ItemStack getProduct() {
+    		return product;
+    	}
+    	
+    	/**
+    	 * Set the price of this trade
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+    	 */
+    	public void setPrice(ItemStack... price) {
+    		Validate.isTrue(price.length <= 2, "The price can only be a maximum of two items");
+    		
+    		this.price = price;
+    	}
+    	
+    	/**
+    	 * Set the product that this villager trades
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+    	 * 
+    	 * @param product the product the villager trades
+    	 */
+    	public void setProduct(ItemStack product) {
+    		this.product = product;
+    	}
+    	
+    	/**
+    	 * Gets the amount of trades this specific trade has left
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+    	 * 
+    	 * @return the amount of trades left
+    	 */
+    	public int getTradesLeft() {
+    		return tradesLeft;
+    	}
+    	
+    	/**
+    	 * Sets the number of trades left for this trade
+    	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Spaces are evil.

File: src/main/java/org/bukkit/inventory/MerchantInventory.java [hunk 27]
```diff
@@ -1,4 +1,13 @@
 package org.bukkit.inventory;
 
+import org.bukkit.entity.Merchant;
+
 public interface MerchantInventory extends Inventory {
+	
+	/**
+	 * Get the merchant that this inventory is registered to
+	 *
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/inventory/MerchantInventory.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Villager.java [hunk 28]
```diff
@@ -3,7 +3,8 @@
 /**
  * Represents a villager NPC
  */
-public interface Villager extends Ageable, NPC {
+public interface Villager extends Ageable, NPC, Merchant {
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Villager.java]: Spaces are evil.

File: src/main/java/org/bukkit/inventory/MerchantInventory.java [hunk 29]
```diff
@@ -1,4 +1,13 @@
 package org.bukkit.inventory;
 
+import org.bukkit.entity.Merchant;
+
 public interface MerchantInventory extends Inventory {
+
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/inventory/MerchantInventory.java]: Spaces are evil.

File: src/main/java/org/bukkit/entity/Merchant.java [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+    	public ItemStack getProduct() {
+    		return product;
+    	}
+    	
+    	/**
+    	 * Set the price of this trade
+    	 * 
+    	 * @param price the price of the trade, a maximum of two items
+    	 */
+    	public void setPrice(ItemStack... price) {
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/entity/Merchant.java]: Needs a `throws IllegalArgumentException` in the Javadoc and a `throws` on the end of the method

**Comment by ase34** [comment]
This will be useful! I hope this PR will be accepted!

**Comment by riking** [comment]
The core team has stated that this does not feel like an appropriate way to approach villager trading; one problem is that the idea of a "Merchant" and its purpose is very ill-defined.

**Pull request closed without merging**